community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py

```diff
@@ -0,0 +1,183 @@
+Sending data across multiple rack requires sending data through additional network switches. 
+
+Example usage:
+``` my_sbatch.sh
+#SBATCH --ntasks-per-node=8
+#SBATCH --nodes=64
+
+export SLURM_HOSTFILE=$(/slurm/scripts/sort_nodes.py)
+
+srun -l hostname | sort
+```
+"""
+import os
+import subprocess
+import uuid
+from typing import List, Optional, Dict, Callable
+
+def order(paths: List[List[str]]) -> List[str]:
+    """
+    Orders the leafs of the tree in a way that minimizes the sum of distance in between 
+    each pair of neighboring nodes in the resulting order.
+    The resulting order will always start from the first node in the input list.
+    The ordering is "stable" with respect to the input order of the leafs i.e.
+    given a choice between two nodes (identical in other ways) it will select "nodelist-smallest" one.
+    """
+    if not paths: return []
+    # lookup table for node names order
+    name_idx = {p[-1]: i for i, p in enumerate(paths)}
+    class Vert:
+        "Represents a vertex in a *network* tree."
+        def __init__(self, name: str, parent: "Vert"):
+            self.name = name
+            self.parent = parent
+            self.children = {}
+            self.lex_key: int = -1
+
+    # build a tree
+    root = Vert("", None)
+    for path in paths:
+        n = root
+        for v in path:
+            if v not in n.children:
+                n.children[v] = Vert(v, n)
+            n = n.children[v]
+
+    # propagate "lexicographical key" from the leaves to the root
+    def set_lex_key(v: Vert) -> None:
+        if not v.children: # this is a node
+            v.lex_key = name_idx[v.name]
+        else:
+            for u in v.children.values():
+                set_lex_key(u)
+            v.lex_key = min(u.lex_key for u in v.children.values())
+    set_lex_key(root)
+
+    # gather leafs
+    result = []
+    def gather_nodes(v: Vert) -> None:
+        if not v.children: # this is a node
+            result.append(v.name)
+        for u in v.children.values():
+            gather_nodes(u)
+    
+    gather_nodes(root)
+    return result
+
+
+class Instance:
+    def __init__(self, name: str, zone: str, physical_host: Optional[str]):
+        self.name = name
+        self.zone = zone
+        self.physical_host = physical_host
+
+
+def make_path(node_name: str, inst: Optional[Instance]) -> List[str]:
+    if not inst:  # node with unknown instance (e.g. hybrid cluster)
+        return ["unknown", node_name]
+    zone = f"zone_{inst.zone}"
+    if not inst.physical_host:  # node without physical host info (e.g. no placement policy)
+        return [zone, "unknown", node_name]
+
+    assert inst.physical_host.startswith("/"), f"Unexpected physicalHost: {inst.physical_host}"
+    parts = inst.physical_host[1:].split("/")
+    if len(parts) >= 4:
+        return [*parts, node_name]
+    elif len(parts) == 3:
+        return [zone, *parts, node_name]  # add zone
```

Comment:
It is not intuitive to me what is the significance of being >=4 or ==3. Do some physical hosts include zone? I am cautious about the special casing here since I expect physical host may change. At min can you add comment as to what you are checking for here, as you have done for other cases above.

Reply:
Made it more robust by removing assertion about length of `parts`.